engine/robot_logic/obstacle_avoidance.py

```diff
@@ -0,0 +1,212 @@
+    2. Once both sensors are displaying the same values, move forward (boundary following)
+    3. Exits boundary following when conditions met in execute_avoid_obstacle() method
+
+    For now, we assume the side ultrasonic sensor values are from the actual sensors, so changes needed for is_sim
+    '''
+    front_ultrasonic = Ultrasonic(0)
+    rf_ultrasonic = Ultrasonic(1)
+    rb_ultrasonic = Ultrasonic(2)
+    lf_ultrasonic = Ultrasonic(3)
+    # robot should be side_margin_sensor meters away from side obstacle, placeholder
+    side_margin_sensor = 1
+    # robot should be margin_to_front_obstacle meters away from front obstacle, placeholder. Want space to turn
+    margin_to_front_obstacle = 5 + max(robot.robot_state.width, robot.robot_state.length)
+    # When boundary following gap in wall but robot cannot fit (check both side and turn 180 and do boundary following again)
+    if lf_ultrasonic.distance() < side_margin_sensor:
+        curr_heading = robot.robot_state.state[2]
+        desired_heading = curr_heading - math.pi if curr_heading > math.pi else curr_heading + math.pi
+        # error is arbitrary
+        heading_err = 10
+        turn_to_target_heading(robot, desired_heading, heading_err)
+    # if there is an obstacle in front of the robot, turn until there isn't
+    elif front_ultrasonic.distance() < margin_to_front_obstacle:
+        robot.robot_state.motor_controller.spin_motors(0, robot.robot_state.turn_angle)
+    # if there is no obstacle in front of the robot but it's still in avoid obstacle because it's following the boundary
+    else:
+        rf_dist = rf_ultrasonic.distance()
+        rb_dist = rb_ultrasonic.distance()
+        # if the robot is approximately parallel to the robot, go forward
+        if abs(rf_dist - rb_dist) < side_margin_sensor:
+            robot.robot_state.motor_controller.spin_motors(robot.robot_state.move_dist, 0)
+        # otherwise, turn until the robot is parallel
+        else:
+            direction = (rf_dist - rb_dist)/abs(rf_dist - rb_dist)
+            robot.robot_state.motor_controller.spin_motors(0, direction * robot.robot_state.turn_angle)
+        # NOTE: TEST WHETHER THIS ACTUALLY WORKS WHEN TURNING:
+        # IF NOT, MIGHT HAVE TO ADD LOCKS TO FORCE THE ROBOT TO TURN UNTIL IT IS PARALLEL TO ROBOT
+        # THIS INVOLVES A LOT OF SEQUENCE OF MOVES, SO IT MIGHT INVOLVE A LOT OF SEQUENTIAL UNLOCKING
+        # for example, if robot starts off parallel and it turns until parallel again,
+        # lock1 unlocks when robot no longer detects obstacle in front,
+        # then lock2 unlocks if lock1 unlocked and rb ultrasonic < rf ultrasonic.
+        # This is assuming rb ultrasonic > rf ultrasonic when turning ccw.
+        # otherwise, might have to have a condition where robot not moving and just turning
+        # and gets back to the same heading to keep turning
+# TODO: add dp to not traverse to node already traversed/in obstacle
+    
+def track_obstacle(robot):
```

Comment:
I think track_obstacle makes more sense to be in setup because it needs to be setup in setup. Also, it doesn't really have anything directly to do with obstacle avoidance

Reply:
ill add it to set up!

Reply:
done!